feature: Add Caravan wallet format import/export support

vkprogrammer-001 opened the pull request:
### Description
This PR implements support for importing and exporting wallets in the Caravan wallet format as requested in issue #5.

### Changes

- Added `CaravanExport` struct to handle Caravan wallet format
- Implemented methods to export a wallet to Caravan JSON format
- Added functionality to import Caravan wallet configs and convert to BDK descriptors
- Added documentation with usage examples
- Added comprehensive test coverage for all supported multisig types (P2SH, P2WSH, P2SH-P2WSH)

### Implementation Details
- All code is contained in the wallet/export.rs module alongside the existing FullyNoded export functionality
- Support for different address types required by Caravan (P2SH, P2WSH, P2SH-P2WSH)
- Handles the mapping between BDK descriptor format and Caravan's JSON configuration format

### Testing
The implementation includes tests for:
- Exporting BDK wallets to Caravan format
- Importing Caravan configurations to BDK wallets
- All supported address types (P2SH, P2WSH, P2SH-P2WSH)
- Network detection (mainnet/testnet)

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

Closes #5

ValuedMammal reviewed (commented):
I think this will be a good addition to have. After doing a first pass my main concern is the amount string manipulation that's required, and would prefer to use stronger types where possible, for example `Fingerprint`, `DerivationPath`, and `Xpub`.

ValuedMammal commented on wallet/src/wallet/export.rs:
```diff
@@ -337,4 +675,166 @@ mod test {
+        }"#;
+
+        let import = CaravanExport::from_str(json).unwrap();
+        let (external, internal) = import.to_descriptors().unwrap();
```
🔧  This should also check that we can create a bdk `Wallet` from the resulting descriptor strings. In the current implementation I'm not sure this is possible.

ValuedMammal commented on wallet/src/wallet/export.rs:
```diff
@@ -211,6 +274,281 @@ impl FullyNodedExport {
+    pub name: String,
+    /// BIP32 derivation path
+    #[serde(rename = "bip32Path")]
+    pub bip32_path: String,
+    /// Extended public key
+    pub xpub: String,
+    /// Fingerprint of the master key
+    pub xfp: String,
```
These can be types from the `bitcoin::bip32` module, see my comment https://github.com/bitcoindevkit/bdk_wallet/pull/205#pullrequestreview-3178668546.

If the json config expects the `m/` prefix on bip32 paths, I suppose we could write a trivial Serialize impl that does that.

ValuedMammal commented on wallet/src/wallet/export.rs:
```diff
@@ -211,6 +274,281 @@ impl FullyNodedExport {
+    pub name: String,
+    /// Address type (P2SH, P2WSH, P2SH-P2WSH)
+    #[serde(rename = "addressType")]
+    pub address_type: String,
```
`address_type` could be modeled nicely as an enum with three variants.

vkprogrammer-001 replied:
Replaced string-based address handling with a CaravanAddressType enum featuring P2SH, P2WSH, and P2SHWrappedP2WSH variants. Added serde annotations to ensure JSON compatibility. The implementation is now cleaner, more robust, and type-safe.

ValuedMammal commented:
@vkprogrammer-001 Any interest in keeping this PR up to date? If not we can probably find someone to pick it up.

thunderbiscuit commented:
I had been thinking about this in the context of the [multi-keychain wallet](https://github.com/ValuedMammal/multi-keychain-wallet) PRs and the new `KeyRing` type. This sort of stuff is a really good fit for the keyring, because it could simply have a new constructor on the type which passes these backup formats (similar to how we think of doing it for bitcoin core) as an argument, and bam you're done. This can be extended to other export formats too on popular wallets.

vkprogrammer-001 commented on wallet/src/wallet/export.rs:
```diff
@@ -211,6 +274,281 @@ impl FullyNodedExport {
+    /// Extract xpubs and fingerprints from multi descriptor
+    fn extract_xpubs_from_multi<Ctx: ScriptContext>(
+        multi: &miniscript::descriptor::SortedMultiVec<String, Ctx>,
+    ) -> Result<Vec<CaravanExtendedPublicKey>, &'static str> {
```
thanks for the suggestion @ValuedMammal. I have updated `extract_xpubs_from_multi` to work directly with SortedMultiVec<DescriptorPublicKey, Ctx> instead of parsing from strings.

vkprogrammer-001 commented:
> @vkprogrammer-001 Any interest in keeping this PR up to date? If not we can probably find someone to pick it up.

Yes @ValuedMammal , I am interested in completing this project. I have addressed all the comments and made the necessary changes. PTAL and let me know if there are any further changes needed.

ValuedMammal reviewed (commented):
Thank you @vkprogrammer-001

vkprogrammer-001 commented on wallet/src/wallet/export.rs:
```diff
@@ -211,6 +275,424 @@ impl FullyNodedExport {
     }
 }
 
+/// ExtendedPublicKey structure for Caravan wallet format
+#[derive(Debug, Clone)]
+pub struct CaravanExtendedPublicKey {
```
Yes, definitely! Deserialize can still be derived for CaravanExtendedPublicKey. The custom implementation I added ensures proper handling of the "m/" prefix in the JSON format while internally using correct DerivationPath types.

vkprogrammer-001 replied:
You are right about this. Rust-bitcoin can parse a DerivationPath from a string with or without the m/ prefix. I made the changes.

ValuedMammal replied:
In that case I think it would be better to derive `Deserialize` for `CaravanExtendedPublicKey` rather than manually implementing it. And probably also include a `#[serde(rename = "bip32Path")]` attribute.

ValuedMammal commented:
When this is close to being ready would you mind please dropping the merge commit 9f86668f and just rebase the changes onto current master.

vkprogrammer-001 commented on wallet/src/wallet/export.rs:
```diff
@@ -211,6 +275,424 @@ impl FullyNodedExport {
+    /// Caravan supports P2SH, P2WSH, and P2SH-P2WSH multisig wallets.
+    pub fn export_wallet(wallet: &Wallet, name: &str) -> Result<Self, &'static str> {
+        // Get the descriptor and extract information
+        let descriptor_str = wallet
+            .public_descriptor(KeychainKind::External)
```
This part is also done. Now we are public_descriptor directly instead of converting it to a string and re-parsing it.

oleonardolima commented:
Sorry, but it looks like it needs yet another rebase.

ValuedMammal commented:
ACK c624e851a631a7855f378cec2952a6aed573516a

ValuedMammal commented on src/wallet/export.rs:
```diff
@@ -211,6 +293,421 @@ impl FullyNodedExport {
+        Ok(keys)
+    }
+
+    /// Import a wallet from Caravan format using proper Descriptor construction
```
📌 
```suggestion
    /// Convert this [`CaravanExport`] into a pair of descriptor strings.
```

ValuedMammal reviewed (commented):
reACK 1dbe85ad883f61c961545c1e0ac1340db8d83967
